x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/TokenService.java

```diff
@@ -857,13 +878,19 @@ public void onFailure(Exception e) {
             });
         } else {
             final String newUserTokenId = UUIDs.randomBase64UUID();
+            final Version newTokenVersion = getVersionCompatibility();
             final Map<String, Object> updateMap = new HashMap<>();
             updateMap.put("refreshed", true);
             updateMap.put("refresh_time", clock.instant().toEpochMilli());
-            updateMap.put("superseded_by", getTokenDocumentId(newUserTokenId));
+            if (newTokenVersion.onOrAfter(VERSION_TOKENS_INDEX_INTRODUCED)) {
+                updateMap.put("superseded_by", prependVersionAndEncode(newTokenVersion, getTokenDocumentId(newUserTokenId)));
+            } else {
+                updateMap.put("superseded_by", getTokenDocumentId(newUserTokenId));
```

Comment:
I am reluctant on adding a new field to the mapping for this use case. `superseded_by` is an internal thing so I think it's alright to encode multiple values, as long as, as you said, we have a function doing encoding/decoding without breaking the flow.

Now, I understand that token version and document id are two different things to encode together, and you suggest we keep both things in the same category. You suggest we stay on the "document" category. What if we move to "token" category? This way the value in this field would be a "refresh token value", a known data type, and we can reuse the `findTokenFromRefreshToken` method. True, this entails more computations when following these references, but I believe the save in complexity in this code area is worth it. What do you think? I will start to implement this, but please voice your opposition, and also we can always backtrack if we're not satisfied in the end.

Reply:
I went back and forth on this one and I've decided to keep encoding the version in the "superseded_by" field. The reason is that a token doc contains a single access token and possibly a single refresh token, both of which have the same version. Therefore, by a stretch of reason, the token document is versioned, so it is _ok_ to prepend the version to the token document id.
I don't like adding another field to the mapping for such an internal corner case (when a token created before a full upgraded cluster is refreshed). And also I don't like "encoding" the index name in the same field another way, without a version header, because we expose ourselves in the future to complex heuristics.

Lemme know what you think! If you don't like it, we could go with `alias_name` + `/` + `doc_id` , but again, the `prependVersionAndEncode` looks more familiar to me in this circumstance.

Reply:
I have adapted the implementation so that the `superseded_by` field has the text format "<alias_name>|<document_id>" , but I will try to explain my thought process for originally choosing the other alternative.

> But what's the aversion to having another field in the mapping? 

I do not like this approach because this new field would be having a single value. The code would actually be checking if the field exists. I prefer we earmark the existing `superseded_by` value. And I believe the future proof way to mark the value is to prepend the version. If we're ok with fields with single value, then I believe the proper future-proof way to go about it would be to make `superseded_by` an object in the main security mapping, but this would require reindexing.

> Is there any reason you care what about the version of the superseding token, other than needing to know which index it's in?
What happens if the version stored in superseded_by doesn't match the version in the referenced doc? I don't think you care do you?

The code we're discussing was indeed relying that tokens with new versions are stored in the new index. But I don't see this as a problem, and we already rely on this assumption when we receive a token via the API. I was imagining it was ok to delegate to the `createOAuth2Tokens` function the decision of where to store the token based on it's version. And, in `inner_refresh`, storing the version of the created superseding token was enough of an information to know on which index the superseding token resides.

But, since we now store the index name, instead of the version, in the superseded document, I have adapted the implementation such that the index name is passed in explicitly to the `createOAuth2Token`.

Reply:
I hold my preference to have a formatted keyword text field (the alternative proposed here), instead of two text fields.

Firstly, we have text fields with format everywhere, eg the `version` field itself.
The suggested two fields (doc id and index name) would always be used together, and this is a sign they should really be a single field. In the `version` field example we don't store major, minor and patch values as different fields because they are not used independently. Moreover, in this case, the `superseded_by_index` field would always have the same value, so the code would simply check for the existence of a field in the document, which does not sound like an improvement.

Secondly, the two fields should be under the same object type `superseded`, rather than to two fields in the token doc, which is not possible , because mapping for existing fields cannot change. I realize this last one is a somewhat pedantic, weak reason against change, but then you have to weight it against the reason for change.

Reply:
> I hold my preference to have a formatted keyword text field (the alternative proposed here), instead of two text fields.

I'm fine with this